### PR TITLE
refactor: migrate mongodb to instrumentation #250

### PR DESCRIPTION
**Pull request by vmarchaud**
I migrated the plugin to instrumentation class, however i also updated the patch to hook to different files (which are more up to date with latest version and more precise).
It support `mongodb` starting `3.3` which was released in the summer of 2019. I though adding back support for `< 3.3` but i think it make more sense in a seperate PR + see if supporting those old versions is actually needed.

### CODE REVIEW

File: plugins/node/opentelemetry-instrumentation-mongodb/README.md
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/README.md:33]:
```suggestion
    mongodb: { enabled: false, path: '@opentelemetry/plugin-mongodb' }
```

File: plugins/node/opentelemetry-instrumentation-mongodb/README.md
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/README.md:69]:
```suggestion
[devDependencies-url]: https://david-dm.org/open-telemetry/opentelemetry-js-contrib?path=packages%2Fopentelemetry-instrumentation-mongodb&type=dev
```

File: plugins/node/opentelemetry-instrumentation-mongodb/README.md
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/README.md:68]:
```suggestion
[devDependencies-image]: https://david-dm.org/open-telemetry/opentelemetry-js-contrib/dev-status.svg?path=packages/opentelemetry-instrumentation-mongodb
```

File: plugins/node/opentelemetry-instrumentation-mongodb/README.md
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/README.md:67]:
```suggestion
[dependencies-url]: https://david-dm.org/open-telemetry/opentelemetry-js-contrib?path=packages%2Fopentelemetry-instrumentation-mongodb
```

File: plugins/node/opentelemetry-instrumentation-mongodb/README.md
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/README.md:66]:
```suggestion
[dependencies-image]: https://david-dm.org/open-telemetry/opentelemetry-js-contrib/status.svg?path=packages/opentelemetry-instrumentation-mongodb
```

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/package.json:46]:
```suggestion
    "@opentelemetry/tracing": "0.17.0",
```

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/package.json:44]:
```suggestion
    "@opentelemetry/context-async-hooks": "0.17.0",
```

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:65]: Please move the versions array into a const as it is used at more places.

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:165]: Not sure here but I think this is problematic.
Callbacks are optional in mongodb APIs (see e.g. http://mongodb.github.io/node-mongodb-native/3.6/api/Collection.html#remove) and if no callback is given a promise is returned. Looks like we may create a callback here which would change behavior for users.

Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:165]: From what i saw all the internal here are using callback-based system ([here](https://github.com/mongodb/node-mongodb-native/blob/3.6/lib/core/wireprotocol/write_command.js#L9)) so i believe the promise are handled much higher in the API.

File: plugins/node/opentelemetry-instrumentation-mongodb/package.json
Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/package.json:46]: Same as express, not sure why you'd want to pin it ?

**Comment by vmarchaud** [comment]
About [`6d7367b` (#354)](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/354/commits/6d7367b52f05fa3b3f56ef346eaea46101aeddeb): For some obscure reason, tests were failing if `supportedVersions` was a property, i tested with typescript 4.2.2 with the same result. I converted it to a getter which works for some reason.

**Comment by dyladan** [comment]
> About [`6d7367b` (#354)](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/354/commits/6d7367b52f05fa3b3f56ef346eaea46101aeddeb): For some obscure reason, tests were failing if `supportedVersions` was a property, i tested with typescript 4.2.2 with the same result. I converted it to a getter which works for some reason.

Can you paste the output of the broken test?

**Comment by vmarchaud** [comment]
> Can you paste the output of the broken test?

You can see the tests output here: https://github.com/open-telemetry/opentelemetry-js-contrib/runs/1977512692?check_suite_focus=true

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (dyladan) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: There is a built-in protected `supportedVersions` array https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-instrumentation/src/types.ts#L53

Is there a reason you chose to do it manually in the `init`?

Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: What do you mean by manually ? I believe we give it to the constructor for every instrumentation

Comment (dyladan) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: bah i'm sorry i got my wires crossed

Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: in fact the `supportedVersions` might be wrong if instrumentation patches multiple packages for example pg and pg-pool, http and https, so maybe we should remove this optional param anyway and then add version only when you create a new `InstrumentationNodeModuleDefinition` the same you do with name "mongodb" ?

Comment (dyladan) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: i'm fine with that but it might be confusing for the average instrumentation author who isn't a JS maintainer and is looking for the simplest way to patch a single package

Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: instead of
```
new InstrumentationNodeModuleDefinition<typeof mongodb>(
        'mongodb',
        this._supportedVersions,
        undefined,
```
do
```
new InstrumentationNodeModuleDefinition<typeof mongodb>(
        'mongodb',
        ['>=3.3 <4'],
        undefined,

```
I think that is simple enough ?

Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: @obecny I agree that we could remove it from the instrumentation interface.

> I think that is simple enough ?

I did that in the first place (and it worked) but i needed it twice so @Flarna suggested to have a single variable to avoid duplication.

Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: yea for version we can just declare some const that not necessary sits on class directly, but the main point is that it doesnt need to be declared as `_supportedVersions`. And because it was declared as optional we could remove it easily right ?

Comment (dyladan) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: We can remove anything at 0.x 🤣 

I think it is a relatively painless change for what we have so far.

Comment (Flarna) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: I haven't really understood why this fails but to me it looks like `InstrumentationNodeModuleDefinition` should copy the array instead of storing a reference.

I tried to reproduce this locally but failed. plugin-mongodb tests run fine so it doesn't seem to be a general setup issue on my machine.

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:64]: This doesn't work on windows because of the `/`. Using `\\` works (but most likely breaks non windows).

Comment (dyladan) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:64]: Have you tested this on windows? I think this uses the node module resolver which I thought normalized paths.

Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:64]: Even though i think its handled by node, if thats an actual bug we should fix this directly in the instrumentation base package because the mongodb plugin isn't the only one to use internals require

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: After re-reading this it might comes from the fact that i was using a private property even though a readonly is exposed on the abstract class ? I will try tomorrow but anyway that should not behave and "unset" the property

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (Flarna) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:64]: Yes, tested on windows.

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (Flarna) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: It's not caused by the name. The problem is the sequence of calling `super` in constructor and initializing the properties of the derived class. `InstrumentationBase` calls `this.init()`.
See the generated JS code then it's clear why it's `undefined`:
```js
constructor(_config = {}) {
  super('@opentelemetry/instrumentation-mongodb', version_1.VERSION, _config);
  this._config = _config;
  this._supportedVersionsX = ['>=3.3 <4'];
}
```
It's similar as in C++ - never call virtual functions in a constructor...
In contrast to C++ JS calls the correct virtual function but on an object which is not completely constructed.

Comment (dyladan) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: We should add windows to the testing matrix

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:64]: @Flarna Could you open an issue on the main repo with the details ? we should fix this before RC of SDK for sure

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: > It's not caused by the name. The problem is the sequence of calling super in constructor and initializing the properties of the derived class. InstrumentationBase calls this.init().

Right i see, i will fix this then

Comment (Flarna) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:74]: created https://github.com/open-telemetry/opentelemetry-js/issues/1989 as this seems to be an issue in instrumentation module.

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (Flarna) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:64]: https://github.com/open-telemetry/opentelemetry-js/issues/1990

**Comment by dyladan** [comment]
Can you document in the readme and in the package.json that this package is mac/linux only for now until the instrumentation package is updated?

**Comment by vmarchaud** [comment]
> Can you document in the readme and in the package.json that this package is mac/linux only for now until the instrumentation package is updated?

I think it would be better to make a dedicated PR for that, the graphql and grpc instrumentation use internal require for sure and i'm not sure about others. WDYT ?

**Comment by Flarna** [comment]
> I think it would be better to make a dedicated PR for that, the graphql and grpc instrumentation use internal require for sure and i'm not sure about others. WDYT ?

I can confirm that grpc instrumenation tests fail on windows because of the slashes issue.

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (Flarna) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:66]: What about using `path.normalize('mongodb/lib/core/wireprotocol/index.js'),` for now? The turnaround time for a fix in core until it arrives in contrib is not that short.

Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:66]: I'm fine doing that in another PR so we can do one PR for all contrib, WDYT ?

**Review by Flarna — APPROVED** [review]

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:66]: I think we should make a fix in core repo when `InstrumentationNodeModuleFile` is being created us path.normalize ass @Flarna mentioned

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (obecny) [inline plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:46]: Should we still support version `2.x` as previously we did. It is quite old though (older than 3 years) and what about version  >=3.0.0 <=3.2.x ?
Is this something we should really worry about or we are can drop it ?

**Review by obecny — COMMENTED** [review]
generally looks fine, I have just one concern with dropping support for older version we did support with plugin

File: plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts
Comment (vmarchaud) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:46]: I commented on the PR description that IMO it seems fine and if its requested we could re-add support, WDYT ?

Comment (obecny) [reply plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts:46]: sry I missed that, not a blocker

**Review by obecny — APPROVED** [review]

**Comment by vmarchaud** [comment]
Please don't merge the PR as i need to bump api dependency to 0.18

**Review by dyladan — APPROVED** [review]

**Pull request merged**
